planner/core/logical_plans.go

```diff
@@ -375,21 +375,62 @@ func (p *LogicalJoin) GetPotentialPartitionKeys() (leftKeys, rightKeys []*proper
 	return
 }
 
-func (p *LogicalJoin) columnSubstitute(schema *expression.Schema, exprs []expression.Expression) {
-	for i, cond := range p.LeftConditions {
-		p.LeftConditions[i] = expression.ColumnSubstitute(cond, schema, exprs)
+// columnSubstituteAll is used in projection elimination in apply de-correlation.
+// once partial args in the condition can't be substituted by new exprs, the old projection must be remained.
+// eg: projection: constant("guo") --> column8, once upper layer substitution failed here, the lower layer behind
+// projection can't supply column8 anymore.
+//
+//	upper OP (depend on column8)   --> projection(constant "guo" --> column8)  --> lower layer OP
+//	          |                                                       ^
+//	          +-------------------------------------------------------+
+//
+//	upper OP (depend on column8)   --> lower layer OP
+//	          |                             ^
+//	          +-----------------------------+      // Fail: lower layer can't supply column8 anymore.
+func (p *LogicalJoin) columnSubstituteAll(schema *expression.Schema, exprs []expression.Expression) (hasFallback bool) {
+	// make a copy of exprs for convenience of substitution (may change/partially change the expr tree)
+	cpLeftConditions := p.LeftConditions.Clone()
+	cpRightConditions := p.RightConditions.Clone()
+	cpOtherConditions := p.OtherConditions.Clone()
+	cpEqualConditions := make([]*expression.ScalarFunction, len(p.EqualConditions))
```

Comment:
The `columnSubstituteImpl` doesn't modify the input arg in-place then return it as the return value.
So I think we don't need to clone it here.
Just making a slice and calling builtin `copy()` is enough

Reply:
make sense！columnSubstituteImpl only builds a new one from the bottom up when the lower args have changed.